### PR TITLE
delay activation of handshake key until sending / receiving EOED

### LINKED ISSUE #175: delay passing handshake read key to QUIC until receiving EOED
Otherwise, a QUIC stack could pass ClientFinished to picotls before EOED, which would terminate the handshake due to delivery of an unexpected message.

Note that our strategy is to delay passing of the keys until they are ready to be used (i.e. 1-RTT read key is passed to QUIC after receiving ClientFinished, even though it is known to the TLS stack at an earlier moment); the change proposed in this issue aligns with that.

Fixes https://github.com/quicwg/base-drafts/issues/1651.

### PR DESCRIPTION
**Pull request by kazuho**
fixes #175

### CODE REVIEW

**Comment by kazuho** [comment]
@huitema I believe this PR resolves the issue you raised in 
https://github.com/quicwg/base-drafts/issues/1651. Would you mind checking?

**Comment by huitema** [comment]
I would need some time to actually read the code. A bit too late now. Will try tomorrow.

**Comment by kazuho** [comment]
@huitema :+1:

With the commit above, the PR also removes the code that installed the handshake *write* key at the earliest moment on the client side (i.e. when receiving EncryptedExtensions). The update code now installs the key right after EOED is being emitted.

There is a slight chance that this change will break existing QUIC stacks, however I would like to note that in such cases those stacks need to be fixed, because a stack needs to wait closing 0-RTT until it receives ServerFinished; it should not have been relying on the current timing of picotls delivering the handshake write key.

**Comment by huitema** [comment]
Currently, the keys are all passed as a callback triggered by a call to ptls_handle_message(). As long as this remains true, the code should not break.

I am not sure what you mean by delaying install of the handshake write key after EOED is emitted. In the most cases, EOED and Finished are returned from the same call to ptls_handle_message(), after the QUIC stack passes the server second flight to the TLS stack. I would really like to keep it that way.

There is also a potential issue if the server second flight is very long, say a large certificate chain. The client will need to send acknowledgements on epoch 2, as well as maybe Challenge Responses. It needs the epoch 2 write key in order to do that. 

Of course, the client can send the EOED at any time. In fact, it could send it just after the client Hello. And it could still continue sending 0-RTT packet until the 1-RTT write key is available. Is that where we are going?

**Comment by huitema** [comment]
Please don't check in these changes. The Quic server needs the handshake receive key at the same time as the handshake write key, because it has to process acks and retransmissions. Using the key provision to signal end of epochs only looks like a good idea, but it has unwanted side effects. It would be much better to have an independent signaling that an epoch is complete.

File: lib/picotls.c [hunk 1]
```diff
@@ -1129,7 +1126,7 @@ static int retire_early_data_secret(ptls_t *tls, int is_enc)
     free(tls->early_data);
     tls->early_data = NULL;
 
-    return setup_traffic_protection(tls, is_enc, NULL, 2, 1);
+    return setup_traffic_protection(tls, is_enc, NULL, 2);
 }
```
Comment (huitema) [inline lib/picotls.c]: That will break the Quic stacks if the server flight has more than 3 packets. The server is required to send a Challenge in Quic, but without a read key the server will never be able to read the Challenge response. Please revert.

**Comment by kazuho** [comment]
@huitema You are correct. Thank you for noticing the problem.

**Pull request closed without merging**